### PR TITLE
3 part gids with source, layer, id

### LINKED ISSUE #238: Inadequate data to construct request to /place endpoint with data from /search
We should normalize the way we mint our IDs (or at least display them to users) so that the `id` field of a geocode response can be used to retrieve a complete place record on the `/place` endpoint.

Steps to reproduce:
- Search for `http://pelias.bigdev.mapzen.com/v1/search?text=hackney&size=1'
 - get `id : '937569400"' as document ID. I know more than the average user, and know how we used to construct these. I get back `source : "osm"`.
- Try to retrieve document at: `http://pelias.bigdev.mapzen.com/v1/place?id=937569400`. Fails.
 - Gets non-GeocodeJSON error: ```{error: "invalid: must be of the format type:id for ex: 'geoname:4163334'"}'''
- Try to retrieve document at: `http://pelias.bigdev.mapzen.com/v1/place?id=osm:937569400`
 - Gets non-GeocodeJSON error: ```{error: "osm is invalid. It must be one of these values - [geoname, osmnode, osmway, admin0, admin1, admin2, neighborhood, locality, local_admin, osmaddress, openaddresses]"}```

### LINKED ISSUE #277: Use 3-part gids in /place and response bodies

### LINKED ISSUE #284: Inconsistent prefixes for place ids, use source instead and/or return full id in search
Hi there,

When making a call to the `place` endpoint with `ids`, it asks for all ids to be prefixed with a some kind of type and returns this error otherwise:

> [whatever id] is invalid. It must be one of these values - [geoname, osmnode, osmway, admin0, admin1, admin2, neighborhood, locality, local_admin, osmaddress, openaddresses]

The problem is that some of these values don't correspond to anything in the results from the `search` or `reverse` endpoints.

For the `country` layer, the place endpoint expects `admin0`, for `region`, it expects `admin1` and so on. Which can certainly be mapped by the client, be may as well be done by the server.

Also, the British spelling of `neighbourhood` is used by `search` vs. the American spelling (`neighborhood`) for `place`, and `localadmin` in the first becomes `local_admin` in the second. 

Here are a few example results for quattroshapes:

```
{
    "layer": "country",
    "name": "New Zealand",
    "country": "New Zealand",
    "label": "New Zealand",
    "source": "qs",
    "country_a": "NZL",
    "id": "315:adm0:nz:nzl:_"
}
```

```
{
    "layer": "neighbourhood",
    "name": "Parnell",
    "country": "New Zealand",
    "region": "Auckland Region",
    "label": "Parnell, Auckland, Auckland Region",
    "county": "Auckland",
    "source": "qs",
    "country_a": "NZL",
    "id": "2184:_:_:nzl:parnell"
}
```

The `place` endpoint expects these ids:
* `admin0:315:adm0:nz:nzl:_`
* `neighborhood:2184:_:_:nzl:parnell`

Assuming all ids are unique per source, could the source be used as a prefix instead? More generally, could the properties returned when searching or reverse geo-coding contain the exact id that's expected by the `place` endpoint?

Thanks,
Thierry

### PR DESCRIPTION
**Pull request by orangejulius**
Add a new property `gid` to all GeoJSON objects, and changes the expected format of `ids` on the `/place` endpoint to match this GID.

This also moves nearly all of the source/type/layer mapping code into one new helper module.

One interesting change: geonames contains countries, cities, etc, but was not previously selected by the corresponding layers tags. Now the mapping from type to layer in the geojsonify helper class matches the mappings from layer to type (previously in `query/layers.js` but now in `helper/type_mapping.js`)

Another interesting change: it turns out it's not possible to uniquely determine an Elasticsearch type just from a source and a layer! In particular, because both `osmway` and `osmnode` can really be anything, we now have to search across multiple types in the place endpoint.

While on the surface this change is simple, both the goofy geonames mapping and the osmway/node issue are a bit of a challenge to wrap one's head around (at least they were for me). I'm happy to try to explain any of this to anyone or be told of places where something this code does doesn't make sense.

Fixes #277
Fixes #284
Fixes #238

### CODE REVIEW

**Comment by riordan** [comment]
so much patch in this patch

On 9/21/15 19:50, Julian Simioni wrote:
>
> Add a new property |gid| to all GeoJSON objects, and changes the
> expected format of |ids| on the |/place| endpoint to match this GID.
>
> This also moves nearly all of the source/type/layer mapping code into
> one new helper module.
>
> One interesting change: geonames contains countries, cities, etc, but
> was not previously selected by the corresponding layers tags. Now the
> mapping from type to layer in the geojsonify helper class matches the
> mappings from layer to type (previously in |query/layers.js| but now
> in |helper/type_mapping.js|)
>
> Another interesting change: it turns out it's not possible to uniquely
> determine an Elasticsearch type just from a source and a layer! In
> particular, because both |osmway| and |osmnode| can really be
> anything, we now have to search across multiple types in the place
> endpoint.
>
> While on the surface this change is simple, both the goofy geonames
> mapping and the osmway/node issue are a bit of a challenge to wrap
> one's head around (at least they were for me). I'm happy to try to
> explain any of this to anyone or be told of places where something
> this code does doesn't make sense.
>
> Fixes #277 <https://github.com/pelias/api/issues/277>
> Fixes #284 <https://github.com/pelias/api/issues/284>
>
> ------------------------------------------------------------------------
>
>
>         You can view, comment on, or merge this pull request online at:
>
>   https://github.com/pelias/api/pull/292
>
>
>         Commit Summary
>
>   * Remove unused test file
>   * Send 3 part gid (source, layer, id) in GeoJSON responses
>   * Add type list, and raw mappings to/from source/layer and type
>   * Remove all mappings other than helper/type_mapping
>   * Separate real layers and alias layers into separate objects
>   * Extract code to sanitize single id to separate function
>   * Add geoname type to most layers
>   * Leave only geonames specific mapping in gejsonify helper
>   * Fix old reference to /doc endpoint
>   * Expect an array of types from _ids sanitiser
>   * Add mapping function to get type from source and layer
>   * Fix missing space in error message
>   * Derive type, source, and layer list instead of hardcoding
>   * Switch to source:layer:id ids format in /place
>
>
>         File Changes
>
>   * *M* controller/place.js
>     <https://github.com/pelias/api/pull/292/files#diff-0> (2)
>   * *M* helper/geojsonify.js
>     <https://github.com/pelias/api/pull/292/files#diff-1> (79)
>   * *M* helper/query_parser.js
>     <https://github.com/pelias/api/pull/292/files#diff-2> (4)
>   * *A* helper/type_mapping.js
>     <https://github.com/pelias/api/pull/292/files#diff-3> (95)
>   * *M* helper/types.js
>     <https://github.com/pelias/api/pull/292/files#diff-4> (6)
>   * *D* query/layers.js
>     <https://github.com/pelias/api/pull/292/files#diff-5> (15)
>   * *D* query/sources.js
>     <https://github.com/pelias/api/pull/292/files#diff-6> (14)
>   * *D* query/types.js
>     <https://github.com/pelias/api/pull/292/files#diff-7> (16)
>   * *M* sanitiser/_ids.js
>     <https://github.com/pelias/api/pull/292/files#diff-8> (70)
>   * *M* sanitiser/reverse.js
>     <https://github.com/pelias/api/pull/292/files#diff-9> (5)
>   * *M* sanitiser/search.js
>     <https://github.com/pelias/api/pull/292/files#diff-10> (5)
>   * *M* test/ciao/place/basic_place.coffee
>     <https://github.com/pelias/api/pull/292/files#diff-11> (4)
>   * *M* test/ciao/reverse/layers_invalid.coffee
>     <https://github.com/pelias/api/pull/292/files#diff-12> (4)
>   * *M* test/ciao/reverse/layers_mix_invalid_valid.coffee
>     <https://github.com/pelias/api/pull/292/files#diff-13> (4)
>   * *M* test/ciao/reverse/layers_multiple.coffee
>     <https://github.com/pelias/api/pull/292/files#diff-14> (4)
>   * *M* test/ciao/reverse/layers_single.coffee
>     <https://github.com/pelias/api/pull/292/files#diff-15> (4)
>   * *M* test/ciao/reverse/sources_layers_invalid_combo.coffee
>     <https://github.com/pelias/api/pull/292/files#diff-16> (4)
>   * *M* test/ciao/reverse/sources_layers_valid_combo.coffee
>     <https://github.com/pelias/api/pull/292/files#diff-17> (4)
>   * *M* test/ciao/search/layers_alias_address.coffee
>     <https://github.com/pelias/api/pull/292/files#diff-18> (4)
>   * *M* test/ciao/search/layers_invalid.coffee
>     <https://github.com/pelias/api/pull/292/files#diff-19> (4)
>   * *M* test/ciao/search/layers_mix_invalid_valid.coffee
>     <https://github.com/pelias/api/pull/292/files#diff-20> (4)
>   * *M* test/ciao/search/layers_multiple.coffee
>     <https://github.com/pelias/api/pull/292/files#diff-21> (4)
>   * *M* test/ciao/search/layers_single.coffee
>     <https://github.com/pelias/api/pull/292/files#diff-22> (4)
>   * *M* test/ciao/search/sources_layers_invalid_combo.coffee
>     <https://github.com/pelias/api/pull/292/files#diff-23> (4)
>   * *M* test/ciao/search/sources_layers_valid_combo.coffee
>     <https://github.com/pelias/api/pull/292/files#diff-24> (4)
>   * *M* test/unit/controller/place.js
>     <https://github.com/pelias/api/pull/292/files#diff-25> (10)
>   * *M* test/unit/helper/geojsonify.js
>     <https://github.com/pelias/api/pull/292/files#diff-26> (3)
>   * *M* test/unit/helper/query_parser.js
>     <https://github.com/pelias/api/pull/292/files#diff-27> (5)
>   * *A* test/unit/helper/type_mapping.js
>     <https://github.com/pelias/api/pull/292/files#diff-28> (52)
>   * *M* test/unit/mock/backend.js
>     <https://github.com/pelias/api/pull/292/files#diff-29> (2)
>   * *D* test/unit/query/types.js
>     <https://github.com/pelias/api/pull/292/files#diff-30> (23)
>   * *M* test/unit/run.js
>     <https://github.com/pelias/api/pull/292/files#diff-31> (2)
>   * *M* test/unit/sanitiser/_ids.js
>     <https://github.com/pelias/api/pull/292/files#diff-32> (80)
>   * *M* test/unit/sanitiser/_layers.js
>     <https://github.com/pelias/api/pull/292/files#diff-33> (7)
>   * *D* test/unit/sanitiser/_source.js
>     <https://github.com/pelias/api/pull/292/files#diff-34> (127)
>   * *M* test/unit/sanitiser/_sources.js
>     <https://github.com/pelias/api/pull/292/files#diff-35> (3)
>   * *M* test/unit/sanitiser/place.js
>     <https://github.com/pelias/api/pull/292/files#diff-36> (12)
>
>
>         Patch Links:
>
>   * https://github.com/pelias/api/pull/292.patch
>   * https://github.com/pelias/api/pull/292.diff
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pelias/api/pull/292>.
>

-- 
-- 
David Riordan | Product Manager - Search | Mapzen
dave.riordan@mapzen.com | @riordan | gpg 235D 9DC9 5EF6 277C
https://mapzen.com | @mapzen

File: controller/place.js [hunk 1]
```diff
@@ -16,7 +16,7 @@ function setup( backend ){
     var query = req.clean.ids.map( function(id) {
       return {
         _index: 'pelias',
-        _type: id.type,
+        type: id.types,
```
Comment (dianashk) [inline controller/place.js]: Probably should leave a comment here for "future us"

File: helper/type_mapping.js [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+ * derive the list of types, sources, and layers from above mappings
+ */
+var TYPES   = Object.keys(TYPE_TO_SOURCE);
+var SOURCES = Object.keys(SOURCE_TO_TYPE);
+var LAYERS  = Object.keys(LAYER_TO_TYPE);
+
+/**
+ * Calculate the set-style intersection of two arrays
+ */
+var intersection = function intersection(set1, set2) {
```
Comment (dianashk) [inline helper/type_mapping.js]: there is an `intersection` function in `lodash`

File: sanitiser/_ids.js [hunk 3]
```diff
@@ -1,19 +1,58 @@
...
-// id generally looks like 'geoname:4163334' (type:id)
-// so, both type and id are required fields.
+// validate inputs, convert types and apply defaults id generally looks like
+// 'geonames:venue:4163334' (source:layer:id) so, all three are required
 
 var lengthError = 'invalid param \'ids\': length must be >0';
 
 var formatError = function(input) {
-  return 'id `' + input + 'is invalid: must be of the format type:id for ex: \'geoname:4163334\'';
+  return 'id `' + input + ' is invalid: must be of the format type:id for ex: \'geonames:venue:4163334\'';
```
Comment (dianashk) [inline sanitiser/_ids.js]: The message is not right. Says it "must be of the format __type:id__" but should be __source:layer:id__

File: helper/geojsonify.js [hunk 4]
```diff
@@ -22,54 +24,30 @@ var DETAILS_PROPS = [
...
-        if (src.category.indexOf('admin:admin1') !== -1) { return 'region'; }
-        if (src.category.indexOf('admin:admin2') !== -1) { return 'county'; }
-        return 'neighbourhood'; // this could also be 'local_admin'
-      }
-
-      if (src.name) { return 'venue'; }
-      if (src.address) { return 'address'; }
+  if (src._type === 'geoname') {
+    if (src.category && src.category.indexOf('admin') !== -1) {
+      if (src.category.indexOf('admin:city') !== -1) { return 'locality'; }
```
Comment (dianashk) [inline helper/geojsonify.js]: let's use _.contains here instead

Comment (orangejulius) [inline helper/geojsonify.js]: Doh. I actually fully intended to do that (I even brought in lodash and used _.contains a few lines lower.

File: sanitiser/_ids.js (hunk 3 above)
Comment (orangejulius) [inline sanitiser/_ids.js]: ahh good catch

File: helper/type_mapping.js (hunk 2 above)
Comment (orangejulius) [inline helper/type_mapping.js]: oh :trophy:

File: controller/place.js (hunk 1 above)
Comment (orangejulius) [inline controller/place.js]: ah good idea. also, does anyone know what happens differently in elasticsearch for _type vs type? maybe @missinglink does

Comment (orangejulius) [inline controller/place.js]: Ok, looking at [this question on the ES discourse page](https://discuss.elastic.co/t/should-filtering-by-type-and-type-specific-search-url-return-different-results/7008), it looks like the difference is that `_type` is used by many wrappers (like the JS one we use) to change the actual URL, which can obviously only filter against one type. `type`, meanwhile, actually adds type to part of the query, and Elasticsearch will accept an array of possible values to match against no problem.

**Comment by dianashk** [comment]
Clear for takeoff :airplane:

**Pull request merged**